docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,508 @@
+  - topLevelTargetRefKind: Mesh
+    policyName: aaa
+    toTargetRefKind: Mesh
+    conf: $conf2
+  - topLevelTargetRefKind: Mesh
+    policyName: bbb
+    toTargetRefKind: MeshService
+    conf: $conf1
+```
+
+and we'll sort this list based on fields in the following order:
+
+1. topLevelTargetRefKind
+2. toTargetRefKind
+3. policyName
+
+Note: today we basically do 1 and 3, ignoring the field 2.
+
+#### Migration
+
+ResourceRule is a new structure, it will be used when configuring outbounds for the real MeshService resources.
+That's why we don't need backwards compatibility for `to[]`.
+
+For `from[]` policies the sorting algorithm should be changed as well. 
+But the problem is less visible due to the fact `from[].targetRef.kind` is `Mesh` in the overwhelming majority of cases.
+For MeshTrafficPermission it can be `Mesh` and `MeshSubset`, and technically the behaviour of the following policies is going to change:
+
+```yaml
+kind: MeshTrafficPermission
+name: aaa
+spec:
+  targetRef:
+    kind: Mesh
+  to:
+    - targetRef:
+        kind: Mesh
+      default: 
+        action: Deny
+---
+kind: MeshTrafficPermission
+name: bbb
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: MeshSubset
+        tags:
+          kuma.io/service: backend
+      default: 
+        action: Allow
+# resulting behaviour is 'action: Deny` for all
+```
+
+but it probably means the user already had policies in the obscure state. 
+The change is harmless, and we don't need to provide a migration path for `from[]`.
+
+### Sparse ResourceRules structure
+
+ResourceRule structure is going to have only resources that were specified in policies.
+Given the following resources:
+
+```yaml
+type: MeshService
+mesh: mesh-1
+name: my-service
+labels:
+  a-common-label: foo
+---
+type: MeshService
+mesh: mesh-1
+name: labeled-service
+labels:
+  a-common-label: foo
+---
+type: MeshService
+mesh: mesh-1
+name: other-service
+labels: {}
+---
+type: Policy
+mesh: mesh-1
+spec:
+  targetRef:
+    kind: Mesh
+  to:
+    - targetRef:
+         kind: Mesh
+       conf: $conf1
+---
+type: Policy
+mesh: mesh-1
+spec:
+  targetRef:
+    kind: Mesh
+  to:
+    - targetRef:
+         kind: MeshService
+         labels:
+             a-common-label: foo
+       conf: $conf2
+---
+type: Policy
+mesh: mesh-1
+spec:
+  targetRef:
+    kind: Mesh
+  to:
+    - targetRef:
+         kind: MeshService
+         name: my-service
+       conf: $conf3
+```
+
+Kuma CP should produce the following structure:
+
+```yaml
+resourceRules:
+  meshservice:mesh/mesh-1:name/my-service:
+    conf: merge($conf1, $conf2, $conf3)
+  meshservice:mesh/mesh-1:name/labeled-service:
+    conf: merge($conf1, $conf2)
+  mesh:name/mesh-1:
+    conf: $conf1
+```
+
+Note that the structure doesn't contain `other-service`, but instead it contains entry for `mesh-1`.
+The code that computes confs should handle the absence of `other-service` and replace it with conf of the `mesh-1`.
```

Comment:
Added 4th policy and added `another-common-label: foo` to the `labeled-service`, otherwise it wouldn't affect the ResourceRules. 

> Also finally, will this solve the origin order .i.e: https://github.com/kumahq/kuma/issues/9173

Good point, yes we can solve it in this MADR. I added `description` to the OpenAPI `origin` field saying the order matters.